### PR TITLE
Sword game

### PR DESCRIPTION
**Pull request by jfcarter**
This is a two person game and as such requires at least two testers. One person needs to create the arena through the tablet app (instructions below). Please test this game using the knight avatar https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/ROC/SwordGameAvatars/fst/warrior_Rigged.fst

Click Edit -> Open and run script from URL -> then enter this link to add the app to your tablet 
https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/ROC/ROC_Arena_Creator.js

1. Download ROC app through file ROC_Arena_Creator.js (app named "ROC" with a sword icon)(Icon will be replaced later)
2. open Roc app and see two buttons. "Place Arena" and "Clear Arena".
3. Click place arena and you will see an arena placed in front of you with a sign to its left, two knights to its right, a sign above the two knights, and two sets of sword and rock entites placed at both ends of arena. (sign has texture issue which will be fixed soon)
4. Switch to HMD mode
5. Pick up sword and you will hear a "swish" sound followed by an alert that the game is equipping your armor. You will be equipped with invisible armor, health indicator, and a shield after you hear "armor equipped".
6. Grab your shield and begin to battle.
7. When you are hit you will lose one of the hearts floating above your head as well as one of the hearts on your shield. You should also hear me say the word "Hit".
8. When you are blocked your sword will turn red and you will experience some haptic feedback. Your sword is then disabled for 3 second (you can no longer deal damage to your enemy). You will also hear me say the word blocked
9. Hit the Reset sign with your hand laser to reset the game at any point. You will feel haptic feedback when you do.
10. At the end of each game the winners sword will be replaced with a trophy while the losers sword will be replaced with a rubber chicken
11. When you press the sign above the knights it will open your tablet to the marketplace section where you can equip the knight.

### CODE REVIEW

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 1]
```diff
@@ -0,0 +1,469 @@
+(function() {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: Fix indentation in this file

Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: Remove unnecessary commented-out code

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+// When user click the app button, we'll display our app on the tablet screen
+function onClicked() {
+tablet.gotoWebScreen(APP_URL);
+}
+button.clicked.connect(onClicked);
+
+// Handle the events we're recieving from the web UI
+function onWebEventReceived(event) {
+	print("AI_Navigation_App.js received a web event:" + event);
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: remove print statement - looks like it's code for an old app

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+				x: 12,
+				y: prop.y,
+				z: 12
+			}
+		};
+		Entities.editEntity(arena, resizeArena);
+		//create equip areas (swords in rocks) as well as other objects around the arena
+		createEquipAreas();
+	}
+	else if ((info.type == "click") && (info.data == "Clear Arena")) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: formatting of else-if statements, indentation in the try-catch

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+            z: .0599
+        },
+        name: "ROC-Sword2",
+        modelURL: "https://hifi-content.s3.amazonaws.com/jedon/Assets/broadsword_01.fbx",
+        shapeType: "compound",
+        script: "https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/ROC/ROC_Battle_Simulator.js?" + Date.now()
+    };
+    sword2 = Entities.addEntity(swordProperties2);
+
+    //save positions and rotations of swords while in rocks for when resetting game
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: indentation and spacing here

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+// Here, we write a function called "cleanup" which gets executed when
+// this script stops running. It'll remove the app button from the tablet.
+
+function cleanup() {
+	print("deleting tablet app button.");
+	tablet.removeButton(button);
+}
+Script.scriptEnding.connect(cleanup);
+
+}());
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: newline EOF

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2636/HighFidelity-Beta-PR11220-cce2d1b8cc264a6da05a8a8e4765ab9166d50df4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2636/HighFidelity-Beta-PR11220-cce2d1b8cc264a6da05a8a8e4765ab9166d50df4.exe

**Comment by YuppyPlays** [comment]
Testing.

**Comment by Nex-Pro** [comment]
Yuppy, I am doing this

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2638/HighFidelity-Beta-PR11220-0b57df835012f7595dfa63efd6df8c4c81921f7c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2638/HighFidelity-Beta-PR11220-0b57df835012f7595dfa63efd6df8c4c81921f7c.exe

**Comment by Nex-Pro** [comment]
### FAIL

Tried multiple times to run this build. Have posted maybe 10 splats. Tried all ways to get this up and running.
Eventually gave up.

Updated to dev latest. Ran the script in #7039.

Only the reset sign appeared.  There were several missing models ( green bounding boxes ) 
https://hifi-content.s3.amazonaws.com/jedon/Assets/broadsword_01.fbx
https://hifi-content.s3.amazonaws.com/jedon/Assets/rock.fbx
Checking the Model URLs revealed that they are incorrect. ( no file ) 

Shooting the reset sign spawns the swords from different URL.
https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/ROC/SwordGameAssets/broadsword_01.fbx

These links need to be fixed.

**Comment by Nex-Pro** [comment]
Additionally I had to get the raw links to the scripts. Maybe new contributors should get given a guide.

File: unpublishedScripts/marketplace/SwordBattle/ROC_Knight_Changer.js [hunk 6]
```diff
@@ -19,13 +31,11 @@
...
             originalAvatar = BLUE_KNIGHT;
-        } 
-        else if (RED_KNIGHT === MyAvatar.skeletonModelURL) {
+        } else if (RED_KNIGHT === MyAvatar.skeletonModelURL) {
             MyAvatar.skeletonModelURL = UNDEAD_KNIGHT;
             MyAvatar.resetSize();
             originalAvatar = RED_KNIGHT;
-        }
-        else if(UNDEAD_KNIGHT === MyAvatar.skeletonModelURL) {
+        } else if(UNDEAD_KNIGHT === MyAvatar.skeletonModelURL) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Knight_Changer.js]: space after if

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 7]
```diff
@@ -1,469 +1,476 @@
...
+    			Entities.deleteEntity(arena);
+    			Entities.deleteEntity(sword1);
+    			Entities.deleteEntity(sword2);
+                Entities.deleteEntity(rock1);
+                Entities.deleteEntity(rock2);
+                Entities.deleteEntity(resetArea);
+    			Entities.deleteEntity(sign);
+                Entities.deleteEntity(knightifier);
+                Entities.deleteEntity(gate);
+    		} catch(err) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: space after catch

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 8]
```diff
@@ -1,469 +1,476 @@
...
 
-function getPositionToCreateKnightifier() {
-      var direction = Quat.getFront(MyAvatar.orientation);
-      var distance1 = 10;
-      var position1 = Vec3.sum(MyAvatar.position, Vec3.multiply(direction, distance1));
+    // Helper function that gives us a position right in front of the user.
+    //Two versions. true places arena and false places sign
+    function getPositionToCreateArena() {
+        var direction = Quat.getFront(MyAvatar.orientation);
+    	var distance = 10;
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: indentation

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 9]
```diff
@@ -1,469 +1,476 @@
...
+        var position2 = Vec3.sum(position1, Vec3.multiply(newDirection, distance2));
+
+        return position2;
     }
 
-}
+    // Helper function that gives us a position right in front of the user 
+    //Two versions. true places sword and false places rock
+    function getPositionToCreateRockandSword1(check) {
+    	if(check == true) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: space after if

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 10]
```diff
@@ -1,469 +1,476 @@
...
-  	   var direction = Quat.getFront(MyAvatar.orientation);
-       var distance = 17;
-       var position = Vec3.sum(MyAvatar.position, Vec3.multiply(direction, distance));
-       position.y -= .7;
-       return position;
+
+    // Helper function that gives us a position right in front of the user 
+    //Two versions. true places sword and false places rock
+    function getPositionToCreateRockandSword2(check) {
+        if(check == true) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: space after if

File: unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js [hunk 11]
```diff
@@ -59,16 +71,15 @@
         print("Checking to make sure the message sender is correct");
         print("Sword holder ID = " + data[6]);
         print("My avatar ID = " + MyAvatar.sessionUUID);
-        if((channel == swordChannel) && (MyAvatar.sessionUUID == data[6]))
-        {
+        if((channel == swordChannel) && (MyAvatar.sessionUUID == data[6])) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js]: space after if

File: unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js [hunk 12]
```diff
@@ -185,9 +195,7 @@
             };
             Entities.editEntity(myID, usability);
             Script.setTimeout(Wait, 3000);
-        }
-        //if the sword hits the enemy players body then I do damage
-        else if(((collisionName == "ROC-Body") || (collisionName == "ROC-Head")) && firstHit && (enemyShield != shield)) { 
+        } else if(((collisionName == "ROC-Body") || (collisionName == "ROC-Head")) && firstHit && (enemyShield != shield)) { //if the sword hits the enemy players body then I do damage
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js]: space after if

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2642/HighFidelity-Beta-PR11220-9d9e60f648cdf68722579a40056bf9b4dfc0dbc0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2642/HighFidelity-Beta-PR11220-9d9e60f648cdf68722579a40056bf9b4dfc0dbc0.exe

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 13]
```diff
@@ -55,53 +55,53 @@
...
                 Entities.deleteEntity(resetArea);
-    			Entities.deleteEntity(sign);
+                Entities.deleteEntity(sign);
                 Entities.deleteEntity(knightifier);
                 Entities.deleteEntity(gate);
-    		} catch(err) {
-    			print("No arena to erase");
-    		}
-    	}
+            } catch(err) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: space after catch

File: unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js [hunk 14]
```diff
@@ -417,14 +417,14 @@
     // Helper function that gives us a position right in front of the user 
     //Two versions. true places sword and false places rock
     function getPositionToCreateRockandSword1(check) {
-    	if(check == true) {
-    	    var direction = Quat.getFront(MyAvatar.orientation);
-    	    var distance = 3;
-    	    var position = Vec3.sum(MyAvatar.position, Vec3.multiply(direction, distance));
-    	    position.y += 0;
-    	    return position;
+        if(check == true) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Arena_Creator.js]: space after if

File: unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js [hunk 15]
```diff
@@ -195,7 +195,8 @@
             };
             Entities.editEntity(myID, usability);
             Script.setTimeout(Wait, 3000);
-        } else if(((collisionName == "ROC-Body") || (collisionName == "ROC-Head")) && firstHit && (enemyShield != shield)) { //if the sword hits the enemy players body then I do damage
+        //if the sword hits the enemy players body then I do damage
+        } else if(((collisionName == "ROC-Body") || (collisionName == "ROC-Head")) && firstHit && (enemyShield != shield)) {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js]: space after if

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2647/HighFidelity-Beta-PR11220-508d0822f8749bfdad0d5e630f334825733efd07.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2647/HighFidelity-Beta-PR11220-508d0822f8749bfdad0d5e630f334825733efd07.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2669/HighFidelity-Beta-PR11220-df083fca41f34e71c7d2fb1d8b527e0c340069ee.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2669/HighFidelity-Beta-PR11220-df083fca41f34e71c7d2fb1d8b527e0c340069ee.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2674/HighFidelity-Beta-PR11220-c80df4432cd7ba7c684a6417b79c3916c730d483.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2674/HighFidelity-Beta-PR11220-c80df4432cd7ba7c684a6417b79c3916c730d483.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2693/HighFidelity-Beta-PR11220-ecad041eeb4ae877db942c1a756037575640ede2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2693/HighFidelity-Beta-PR11220-ecad041eeb4ae877db942c1a756037575640ede2.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2707/HighFidelity-Beta-PR11220-a87537e5e94445278c7768e34d443fcc7ab5ce12.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2707/HighFidelity-Beta-PR11220-a87537e5e94445278c7768e34d443fcc7ab5ce12.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2718/HighFidelity-Beta-PR11220-d3e43d8cbfec78f44cbd3411c8a041afb8863210.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2718/HighFidelity-Beta-PR11220-d3e43d8cbfec78f44cbd3411c8a041afb8863210.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2725/HighFidelity-Beta-PR11220-3dd7c0a913b4302d876a83023768a69a9d88493e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2725/HighFidelity-Beta-PR11220-3dd7c0a913b4302d876a83023768a69a9d88493e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2767/HighFidelity-Beta-PR11220-604909b6030953744770c5b133ee900bbfeb0b1b.dmg

**Comment by misslivirose** [comment]
build this please

File: unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js [hunk 16]
```diff
@@ -60,30 +61,39 @@
...
-        equipSound = SoundCache.getSound(Script.resolvePath("./SwordGameSounds/Sword_collide.wav"));
+        swishSound = SoundCache.getSound(Script.resolvePath("./SwordGameSounds/Sword_collide.wav"));
+        equipSound = SoundCache.getSound(Script.resolvePath("./SwordGameSounds/equipSound.wav"));
     }
 
     _this.onReceivedMessage = function(channel, message, senderID) {
-        var data = JSON.parse(message);
         //since its an entity script every player recieves the message so I need to make sure the message is only run for the player with the sword
+        try {
+        var data = JSON.parse(message);
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js]: Indentation

File: unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js [hunk 17]
```diff
@@ -94,10 +104,22 @@
...
+                    visible: true
+                };
                 Entities.editEntity(body, parentBodyArmor);
                 Entities.editEntity(head, parentHeadArmor);
                 Entities.editEntity(focalPoint, parentHeartArmor);
-            }, 1000);
+                Entities.editEntity(shield, visibleShield);
+
+                //Attaching everything
+                Script.setTimeout(function() {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/SwordBattle/ROC_Battle_Simulator.js]: Indentation

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2785/HighFidelity-Beta-PR11220-604909b6030953744770c5b133ee900bbfeb0b1b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2785/HighFidelity-Beta-PR11220-604909b6030953744770c5b133ee900bbfeb0b1b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2796/HighFidelity-Beta-PR11220-7fe664a9cfdaec1ae3ef0e346aae250d198c995c.dmg

**Comment by misslivirose** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2821/HighFidelity-Beta-PR11220-7fe664a9cfdaec1ae3ef0e346aae250d198c995c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2821/HighFidelity-Beta-PR11220-7fe664a9cfdaec1ae3ef0e346aae250d198c995c.dmg

**Comment by Nex-Pro** [comment]
@misslivirose Do you want me to re-test this ?

**Comment by misslivirose** [comment]
@Nex-Pro I think that @jfcarter is making some performance improvements before we retest - thanks for checking!

**Comment by Nex-Pro** [comment]
@misslivirose In which case do you think we should remove the QA Ready label ?

**Comment by misslivirose** [comment]
Closing as this has been moved to new [hifi/hifi-content repo](https://github.com/highfidelity/hifi-content/pull/5).

**Pull request closed without merging**